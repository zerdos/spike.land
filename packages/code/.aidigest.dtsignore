
src/@/components/ui/*
src/@/lib/esbuild*
tsconfig.app.tsbuildinfo

