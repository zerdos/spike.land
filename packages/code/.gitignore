# Package-specific build artifacts
/public/
/monaco-workers/
/dts/

# Package-specific caches and temporary files
/.yarn/
/certs/
/errors.txt

# Generated files
sw2.ts

# Local development
*.local

# Build output
/dist/
/dist-vite/
/.vite/
