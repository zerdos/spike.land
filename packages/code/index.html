<!DOCTYPE html>
<html lang="en">
  <head>
    <base href="/" />
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>spike.land</title>
    <link
      rel="preload"
      href="https://fonts.googleapis.com/css2?family=Roboto+Flex:opsz,wght@8..144,100..1000&display=swap"
      as="style"
    />
    <link
      href="https://fonts.googleapis.com/css2?family=Roboto+Flex:opsz,wght@8..144,100..1000&display=swap"
      rel="stylesheet"
    />

    <script type="importmap">
      {
        "imports": {
          "@/": "/@/",
          "@emotion/react/jsx-runtime": "/emotionJsxRuntime.mjs",
          "@emotion/react/jsx-dev-runtime": "/emotionJsxRuntime.mjs",
          "@emotion/styled": "/emotionStyled.mjs", 
          "react/jsx-runtime": "/jsx.mjs",
          "react-dom/server": "/reactDomServer.mjs",
          "react-dom/client": "/reactDomClient.mjs",
          "@emotion/react": "/emotion.mjs",
          "react": "/reactMod.mjs",
          "framer-motion": "/@/workers/framer-motion.mjs",
          "react-dom": "/reactDom.mjs"
        }
      }
    </script>
  </head>

  <body>
    <div id="embed"></div>
    <script type="module" src="./src/start.ts"></script>
  </body>
</html>
