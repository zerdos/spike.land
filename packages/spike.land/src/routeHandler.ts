import { createClerkClient } from "@clerk/backend";
import {
  HTML,
  importMap,
  importMapReplace,
  md5,
  sanitizeSession,
  sessionToJSON,
} from "@spike-npm-land/code";

import type { Code } from "./chatRoom";
import type _Env from "./env";

export interface AutoSaveEntry {
  timestamp: number;
  code: string;
}

export class RouteHandler {
  constructor(private code: Code) {}

  async handleRoute(
    request: Request,
    url: URL,
    path: string[],
  ): Promise<Response> {
    const routeHandler = this.getRouteHandler(path[0]);
    return routeHandler
      ? routeHandler(request, url, path)
      : new Response("Not found", { status: 404 });
  }

  private getRouteHandler(
    route: string,
  ): ((req: Request, url: URL, path: string[]) => Promise<Response>) | null {
    const routes: Record<
      string,
      (
        req: Request,
        url: URL,
        path: string[],
      ) => Promise<Response>
    > = {
      users: this.handleUsersRoute.bind(this),
      websocket: this.handleWebsocketRoute.bind(this),
      code: this.handleCodeRoute.bind(this),
      "index.tsx": this.handleCodeRoute.bind(this),
      "session.json": this.handleSessionRoute.bind(this),
      lazy: this.handleLazyRoute.bind(this),
      request: this.handleRequestRoute.bind(this),
      list: this.handleListRoute.bind(this),
      room: this.handleRoomRoute.bind(this),
      path: this.handlePathRoute.bind(this),
      "index.mjs": this.handleJsRoute.bind(this),
      "index.js": this.handleJsRoute.bind(this),
      "index.css": this.handleCssRoute.bind(this),
      "to-string.js": this.handleRenderToStr.bind(this),
      "wrapper.js": this.handleWrapRoute.bind(this),
      "wrapped": this.handleWrapHTMLRoute.bind(this),
      js: this.handleJsRoute.bind(this),
      htm: this.handleHtmlRoute.bind(this),
      env: this.handleEnvRoute.bind(this),
      screenshot: this.handleScreenShotRoute.bind(this),

      my: this.handleMyCode.bind(this),
      hashCode: this.handleHashCodeRoute.bind(this),
      "": this.handleDefaultRoute.bind(this),

      undefined: this.handleDefaultRoute.bind(this),
      "null": this.handleDefaultRoute.bind(this),
      hydrated: this.handleDefaultRoute.bind(this),
      worker: this.handleDefaultRoute.bind(this),

      dehydrated: this.handleDefaultRoute.bind(this),
      iframe: this.handleDefaultRoute.bind(this),
      embed: this.handleDefaultRoute.bind(this),

      public: this.handleDefaultRoute.bind(this),
      // New routes for auto-save functionality

      // "auto-save": this.handleAutoSaveRoute.bind(this),

      // "history": this.handleCodeHistory.bind(this),
      // New route for serving saved versions
      live: this.handleLiveRoute.bind(this),
    };

    return routes[route] || null;
  }
  // private async handleCodeHistory() {
  //   const history = [] as string[];
  //   return new Response(JSON.stringify(history), {
  //     status: 200,
  //     headers: {
  //       "Access-Control-Allow-Origin": "*",
  //       "Cross-Origin-Embedder-Policy": "require-corp",
  //       "Cache-Control": "no-cache",
  //       "Content-Type": "application/json; charset=UTF-8",
  //     },
  //   });
  // }

  // private async handleAutoSaveRoute(
  //   _request: Request,
  //   url: URL,
  //   path: string[],
  // ): Promise<Response> {
  //   const action = path[1];

  //   try {
  //     switch (action) {
  //       case "history": {
  //         const subAction = path[2];
  //         if (subAction === "delete") {
  //           const itemToDelete = path[3];
  //           const uniqueHistory = await this.getUniqueHistory();
  //           const snapshotToSave = uniqueHistory.filter(
  //             (s) => s.timestamp !== Number(itemToDelete),
  //           );

  //           this.code.setAutoSaveHistory(snapshotToSave);
  //           return new Response("Auto-save history item deleted", {
  //             status: 200,
  //           });
  //         }

  //         return this.getAutoSaveHistory();
  //       }

  //       case "restore": {
  //         const restoreTimestamp = Number(path[2]);

  //         if (Number.isNaN(restoreTimestamp) || !restoreTimestamp) {
  //           return new Response("Failed to restore code: Invalid timestamp", {
  //             status: 400,
  //           });
  //         }

  //         const success = await this.code.restoreFromAutoSave(restoreTimestamp);

  //         if (success) {
  //           return new Response("Code restored successfully", { status: 200 });
  //         } else {
  //           return new Response("Failed to restore code. Snapshot not found.", {
  //             status: 404,
  //           });
  //         }
  //       }
  //       default: {
  //         await this.code.autoSave();
  //         const codeSpace = url.searchParams.get("room");
  //         const { html } = this.code.getSession();
  //         const respText = HTML.replace(
  //           `<script type="importmap"></script>`,
  //           `<script type="importmap">${JSON.stringify(importMap)}</script>`,
  //         ).replace(
  //           `<!-- Inline LINK for initial theme -->`,
  //           `<link rel="preload" href="/live/${codeSpace}/index.css" as="style">
  //            <link rel="stylesheet" href="/live/${codeSpace}/index.css">
  //                   `,
  //         ).replace(
  //           '<div id="embed"></div>',
  //           `<div id="embed">${html}</div>`,
  //         );

  //         const headers = new Headers({
  //           "Access-Control-Allow-Origin": "*",
  //           "Cross-Origin-Embedder-Policy": "require-corp",
  //           "Cross-Origin-Resource-Policy": "cross-origin",
  //           "Cross-Origin-Opener-Policy": "same-origin",
  //           "Cache-Control": "no-cache",
  //           "Content-Encoding": "gzip",
  //           "Content-Type": "text/html; charset=UTF-8",
  //           "content_hash": md5(respText),
  //         });

  //         return new Response(respText, { status: 200, headers });
  //       }
  //     }
  //   } catch (error) {
  //     console.error("Error in handleAutoSaveRoute:", error);
  //     return new Response("Internal server error", { status: 500 });
  //   }
  // }

  // private async getUniqueHistory(): Promise<AutoSaveEntry[]> {
  //   const history = await this.code.getAutoSaveHistory();

  //   const now = Date.now();
  //   const lastHour = now - 1000 * 60 * 60;
  //   const last2days = now - 1000 * 60 * 60 * 24 * 2;

  //   const uniqueMap = new Map<number, AutoSaveEntry>();

  //   for (const entry of history) {
  //     let timestamp = entry.timestamp;

  //     if (timestamp <= lastHour) {
  //       if (timestamp > last2days) {
  //         timestamp = Math.floor(timestamp / 300) * 300;
  //       } else {
  //         timestamp = Math.floor(timestamp / 3000) * 3000;
  //       }
  //     }

  //     // If this timestamp doesn't exist in the map, or if this entry is more recent,
  //     // update the map
  //     if (
  //       !uniqueMap.has(timestamp) ||
  //       entry.timestamp > uniqueMap.get(timestamp)!.timestamp
  //     ) {
  //       uniqueMap.set(timestamp, { ...entry, timestamp });
  //     }
  //   }

  //   const uniqueHistory = Array.from(uniqueMap.values());

  //   // Sort the history in descending order (most recent first)
  //   uniqueHistory.sort((a, b) => b.timestamp - a.timestamp);

  //   // this.code.setAutoSaveHistory(uniqueHistory);
  //   return uniqueHistory;
  // }
  // private async getAutoSaveHistory(): Promise<Response> {
  //   try {
  //     const uniqueHistory = await this.getUniqueHistory();

  //     return new Response(JSON.stringify(uniqueHistory), {
  //       status: 200,
  //       headers: {
  //         "Content-Type": "application/json",
  //         "Access-Control-Allow-Origin": "*",
  //       },
  //     });
  //   } catch (error) {
  //     console.error("Error in getAutoSaveHistory:", error);
  //     return new Response("Failed to retrieve auto-save history", {
  //       status: 500,
  //     });
  //   }
  // }

  private async handleUsersRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    const activeUsers = this.code.wsHandler.getActiveUsers(codeSpace || "");

    return new Response(JSON.stringify({ users: activeUsers }), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Content-Type": "application/json",
        "Cache-Control": "no-cache",
      },
    });
  }

  private async handleWebsocketRoute(request: Request): Promise<Response> {
    if (request.headers.get("Upgrade") !== "websocket") {
      return new Response("Expected websocket", { status: 400 });
    }

    const pair = new WebSocketPair();
    await this.code.wsHandler.handleWebsocketSession(pair[1]);
    return new Response(null, {
      status: 101,
      statusText: "Switching Protocols",
      webSocket: pair[0],
    });
  }

  private async handleCodeRoute(): Promise<Response> {
    return new Response(this.code.getSession().code, {
      status: 200,
      headers: new Headers({
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        content_hash: md5(this.code.getSession().code),
        "Content-Type": "application/javascript; charset=UTF-8",
      }),
    });
  }

  private async handleSessionRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    const body = sessionToJSON(
      sanitizeSession({ ...this.code.getSession(), codeSpace }),
    );
    return new Response(body, {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "Content-Encoding": "gzip",
        "Content-Type": "application/json; charset=UTF-8",
      },
    });
  }

  private async handleLazyRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");

    return new Response(
      `import { jsx as jsX } from "@emotion/react";
       import {LoadRoom} from "/live/lazy/js";
       export default ()=>jsX(LoadRoom, { room:"${codeSpace}"}) ;
       `,
      {
        status: 200,
        headers: {
          "Access-Control-Allow-Origin": "*",
          "Content-Encoding": "gzip",
          "Cross-Origin-Embedder-Policy": "require-corp",
          "Cache-Control": "no-cache",
          "Content-Type": "application/javascript; charset=UTF-8",
        },
      },
    );
  }

  private async handleRequestRoute(request: Request): Promise<Response> {
    return new Response(JSON.stringify({ ...request }), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "Content-Encoding": "gzip",
        "Content-Type": "application/json; charset=UTF-8",
      },
    });
  }

  private async handleListRoute(): Promise<Response> {
    const list = await this.code.getState().storage.list({
      allowConcurrency: true,
    });
    return new Response(JSON.stringify({ ...list }), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Content-Encoding": "gzip",
        "Cache-Control": "no-cache",
        "Content-Type": "application/json; charset=UTF-8",
      },
    });
  }

  private async handleMyCode(request: Request): Promise<Response> {
    const { CLERK_SECRET_KEY: secretKey } = this.code.getEnv();
    const publishableKey = "pk_live_Y2xlcmsuc3Bpa2UubGFuZCQ";

    const clerkClient = createClerkClient({
      secretKey,
      publishableKey,
    });
    const { isSignedIn } = await clerkClient.authenticateRequest(request, {
      jwtKey: `-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAp3A1CStTaorSWlsiSscH
gi2ERdl1KfVsBhvuCIYHlhCowyYgvQhCpQwMD2nkley8WS+Iw8XC8s9yU0S31ONr
mK8zh7e7X/QoCrwQ7SapqTsrg3ryJXWrVeAmG+F4kNvmS6xvyoI+czgzR3gCmE+f
2Ge2cJ6fUQ1hh1jvVUXBdEe8TwRM8zZrlxKJkks3zDjvaPJkJvBqO9Qc52k9i5Sy
0+NnG2ZXuO1Iz7IVB9ow9PkUK/R9+lyB5jASkF2Z8SRksaqJDV+ycEYMd87sO73H
gPWHBSgqBcFixJbT0vLhddwwoqx1pYlnEPlU07NNQHi2JNOQoxsUXJAj/3+w5z6V
hQIDAQAB
-----END PUBLIC KEY-----
`,
    });

    return new Response(JSON.stringify({ isSignedIn }), { status: 200 });
  }
  private async handleDefaultRoute(): Promise<Response> {
    // const url = new URL(r);
    const { html, codeSpace, css } = this.code.getSession();
    const respText = HTML.replace("${JSON.stringify(importMap)}", JSON.stringify(importMap))
      .replaceAll("${codeSpace}", codeSpace).replace("/* criticalCss */", css).replace("${html}",html);

    const headers = new Headers({
      "Access-Control-Allow-Origin": "*",
      "Cross-Origin-Embedder-Policy": "require-corp",
      "Cross-Origin-Resource-Policy": "cross-origin",
      "Cross-Origin-Opener-Policy": "same-origin",
      "Cache-Control": "no-cache",
      "Content-Encoding": "gzip",
      "Content-Type": "text/html; charset=UTF-8",
      "content_hash": md5(respText),
    });

    return new Response(respText, { status: 200, headers });
  }

  private async handleRoomRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    return new Response(JSON.stringify({ codeSpace }), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "Content-Type": "application/json; charset=UTF-8",
      },
    });
  }

  private async handlePathRoute(
    _request: Request,
    _url: URL,
    path: string[],
  ): Promise<Response> {
    return new Response(path.join("----"), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "Content-Type": "application/javascript; charset=UTF-8",
      },
    });
  }

  private async handleLiveRoute(
    _request: Request,
    _url: URL,
    path: string[],
  ): Promise<Response> {
    if (path[3] === "index.tsx" && path[4]) {
      const timestamp = parseInt(path[4]);
      const savedVersion = await this.code.getState().storage.get(
        `savedVersion_${timestamp}`,
      );

      if (savedVersion) {
        return new Response(savedVersion as string, {
          status: 200,
          headers: {
            "Access-Control-Allow-Origin": "*",
            "Cross-Origin-Embedder-Policy": "require-corp",
            "Cache-Control": "no-cache",
            "Content-Type": "application/javascript; charset=UTF-8",
          },
        });
      }
    }

    return new Response("Not found", { status: 404 });
  }

  private async handleWrapHTMLRoute(): Promise<Response> {
    const { html, codeSpace } = this.code.getSession();

    const respText = HTML.replace(
      `<script type="importmap"></script>`,
      `<script type="importmap">${JSON.stringify(importMap)}</script>`,
    )
      .replace(
        `<!-- Inline LINK for initial theme -->`,
        `<!-- Inline LINK for initial theme -->
              <link rel="preload" href="/live/${codeSpace}/index.css" as="style">
              <link rel="stylesheet" href="/live/${codeSpace}/index.css">
        `,
      ).replace(
        '<div id="embed"></div>',
        `<div id="embed">${html}</div>`,
      ).replace("/start.mjs", `https://js.spike.land?codeSpace=${codeSpace}`);

    return new Response(respText, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        content_hash: md5(respText),
        "Content-Type": "text/html; charset=UTF-8",
      },
    });
  }

  private async handleWrapRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    const origin: string = this.code.getOrigin();

    const code = `import App from "${origin}/live/${codeSpace}/index";
    import { renderApp } from "${origin}/@/lib/render-app.mjs";
    
    window.renderedApp = renderApp({ App, rootElement: document.getElementById("embed") });

    `;

    return new Response(code, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "x-typescript-types": this.code.getOrigin() + "/live/index.tsx",
        content_hash: md5(code),
        "Content-Type": "application/javascript; charset=UTF-8",
      },
    });
  }

  private async handleScreenShotRoute(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    const origin: string = this.code.getOrigin();
    return fetch(
      `https://spike-land-renderer.spikeland.workers.dev/?url=${origin}/live/${codeSpace}/embed&now=${Date.now()}`,
    ) as unknown as Promise<Response>;
  }

  private async handleRenderToStr(
    _request: Request,
    url: URL,
  ): Promise<Response> {
    const codeSpace = url.searchParams.get("room");
    const origin: string = this.code.getOrigin();

    const code = `import App from "${origin}/live/${codeSpace}/index";
    import { jsx } from "@emotion/react/jsx-runtime";
     import { renderToString } from "react-dom/server";
    
    const str = renderToString( /*#__PURE__*/_jsx(App, {}));
    
    globalThis.renderedStr = str;


  }
    `;

    return new Response(code, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "x-typescript-types": this.code.getOrigin() + "/live/index.tsx",
        content_hash: md5(code),
        "Content-Type": "application/javascript; charset=UTF-8",
      },
    });
  }

  private async handleHtmlRoute(): Promise<Response> {
    const html = this.code.getSession().html;

    return new Response(html, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        content_hash: md5(html),
        "Content-Type": "text; charset=UTF-8",
      },
    });
  }

  private async handleJsRoute(): Promise<Response> {
    const replaced = importMapReplace(
      this.code.getSession().transpiled,
      this.code.getOrigin(),
    );

    return new Response(replaced, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "x-typescript-types": this.code.getOrigin() + "/live/index.tsx",
        content_hash: md5(replaced),
        "Content-Type": "application/javascript; charset=UTF-8",
      },
    });
  }

  private async handleCssRoute(): Promise<Response> {
    return new Response(this.code.getSession().css, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        content_hash: md5(this.code.getSession().css),
        "Content-Type": "text/css; charset=UTF-8",
      },
    });
  }

  private async handleEnvRoute(): Promise<Response> {
    return new Response(JSON.stringify(this.code.getEnv()), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cache-Control": "no-cache",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Content-Type": "text/html; charset=UTF-8",
      },
    });
  }

  private async handleHashCodeRoute(
    _request: Request,
    _url: URL,
    path: string[],
  ): Promise<Response> {
    const hashCode = String(Number(path[1]));
    const patch = await this.code.getState().storage.get<
      { patch: string; oldHash: number; }
    >(hashCode, {
      allowConcurrency: true,
    });

    return new Response(JSON.stringify(patch || {}), {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Cross-Origin-Embedder-Policy": "require-corp",
        "Cache-Control": "no-cache",
        "Content-Type": "application/json; charset=UTF-8",
      },
    });
  }
}
