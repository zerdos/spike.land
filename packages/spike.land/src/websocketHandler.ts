import { applySessionPatch, computeSessionHash, ICodeSession } from "@spike-npm-land/code";
import type { Code } from "./chatRoom";

export interface WebsocketSession {
  webSocket: WebSocket;
  name?: string;
  quit: boolean;
  subscribedTopics: Set<string>;
  pongReceived: boolean;
  blockedMessages: any[];
}

export class WebSocketHandler {
  private wsSessions: WebsocketSession[] = [];
  private topics = new Map<string, Set<WebSocket>>();
  private code: Code;

  constructor(code: Code) {
    this.code = code;
  }

  getWsSessions() {
    return this.wsSessions;
  }

  setTopics(topics: Map<string, Set<WebSocket>>) {
    this.topics = topics;
  }

  pushToWsSession(session: Partial<ICodeSession>) {
    this.wsSessions.push(session as WebsocketSession);
  }

  handleWebsocketSession(webSocket: WebSocket) {
    webSocket.accept();

    const session: WebsocketSession = {
      webSocket,
      quit: false,
      subscribedTopics: new Set(),
      pongReceived: true,
      blockedMessages: [],
    };

    this.wsSessions.push(session);

    // Send initial handshake
    webSocket.send(JSON.stringify({
      type: "handshake",
      hash: computeSessionHash(this.code.getSession()),
    }));

    // Setup ping interval
    const pingInterval = setInterval(() => {
      if (!session.pongReceived) {
        webSocket.close();
        clearInterval(pingInterval);
        return;
      }
      session.pongReceived = false;
      webSocket.send(JSON.stringify({ type: "ping" }));
    }, 30000);

    // Handle messages
    webSocket.addEventListener("message", (msg: MessageEvent) => {
      this.processWsMessage(msg, session);
    });

    // Handle close
    webSocket.addEventListener("close", () => {
      session.quit = true;
      clearInterval(pingInterval);
      const index = this.wsSessions.indexOf(session);
      if (index > -1) {
        this.wsSessions.splice(index, 1);
      }
    });
  }

  processWsMessage = async (msg: MessageEvent, session: WebsocketSession) => {
    const data = JSON.parse(msg.data as string);

    if (data.type === "ping") {
      session.webSocket.send(JSON.stringify({ type: "pong" }));
      return;
    }

    if (data.type === "pong") {
      session.pongReceived = true;
      return;
    }

    if (data.type === "subscribe") {
      for (const topic of data.topics) {
        session.subscribedTopics.add(topic);
        if (!this.topics.has(topic)) {
          this.topics.set(topic, new Set());
        }
        this.topics.get(topic)?.add(session.webSocket);
      }
      return;
    }

    if (data.type === "unsubscribe") {
      for (const topic of data.topics) {
        session.subscribedTopics.delete(topic);
        this.topics.get(topic)?.delete(session.webSocket);
      }
      return;
    }

    if (data.type === "publish") {
      const subscribers = this.topics.get(data.topic);
      if (subscribers) {
        const message = JSON.stringify({
          type: "message",
          topic: data.topic,
          data: data.data,
        });
        for (const subscriber of subscribers) {
          subscriber.send(message);
        }
      }
      return;
    }

    if (data.patch) {
      try {
        const currentSession = this.code.getSession();
        const currentHash = computeSessionHash(currentSession);

        if (currentHash !== data.oldHash) {
          console.error("Hash mismatch");
          session.webSocket.send(JSON.stringify({
            type: "error",
            message: "Session hash mismatch",
          }));
          return;
        }

        const patchedSession = applySessionPatch(currentSession, data);
        await this.code.updateAndBroadcastSession(patchedSession);

        session.webSocket.send(JSON.stringify({
          type: "ack",
          hashCode: computeSessionHash(patchedSession),
        }));
      } catch (error) {
        console.error("Error applying patch:", error);
        const errorMessage = error instanceof Error ? error.message : "Unknown error";
        session.webSocket.send(JSON.stringify({
          type: "error",
          message: "Failed to apply patch " + errorMessage,
        }));
      }
      return;
    }

    if (data.target) {
      const targetSession = this.wsSessions.find(s => s.name === data.target);
      if (targetSession && targetSession.webSocket.readyState === 1) {
        targetSession.webSocket.send(msg.data);
      }
      return;
    }

    if (data.name) {
      session.name = data.name;
    }
  };

  getActiveUsers(codeSpace: string): string[] {
    return this.wsSessions
      .filter(session => session.subscribedTopics.has(codeSpace))
      .map(session => session.name || "anonymous")
      .filter(Boolean);
  }

  broadcast(message: any, excludeSession?: WebsocketSession) {
    for (const session of this.wsSessions) {
      if (session.webSocket.readyState === 1 && session !== excludeSession) {
        try {
          session.webSocket.send(typeof message === "string" ? message : JSON.stringify(message));
        } catch (error) {
          console.error("Error broadcasting to session:", error);
          session.webSocket.close();
        }
      }
    }
  }
}
