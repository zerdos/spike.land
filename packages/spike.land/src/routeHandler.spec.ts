import { importMap } from "@spike-npm-land/code";
const md5Promise = import("@spike-npm-land/code").then(m => m.md5);
import { beforeEach, describe, expect, it, vi } from "vitest";
import type { Code } from "./chatRoom";
import { RouteHandler } from "./routeHandler";
import { WebSocketHandler } from "./websocketHandler";

vi.mock("snakecase-keys", () => ({}));

describe("RouteHandler", () => {
  let routeHandler: RouteHandler;
  let mockCode: Partial<Code>;

  beforeEach(() => {
    // Create a mock Code object with necessary methods
    mockCode = {
      getSession: vi.fn().mockReturnValue({
        code: "mock code",
        html: "mock html",
        css: "mock css",
        transpiled: "mock transpiled code",
        codeSpace: "test-space",
      }),
      getState: vi.fn().mockReturnValue({
        storage: {
          list: vi.fn().mockResolvedValue({}),
          get: vi.fn().mockResolvedValue(null),
        },
      }),
      getEnv: vi.fn().mockReturnValue({}),
      getOrigin: vi.fn().mockReturnValue("https://example.com"),
      wsHandler: {
        handleWebsocketSession: vi.fn().mockResolvedValue(undefined),
      } as unknown as WebSocketHandler,
    };

    // Create RouteHandler with mock Code
    routeHandler = new RouteHandler(mockCode as Code);
  });

  describe("handleRoute", () => {
    it("should return 404 for unknown route", async () => {
      const request = new Request("https://example.com/unknown");
      const url = new URL("https://example.com/unknown");

      const response = await routeHandler.handleRoute(request, url, ["unknown"]);

      expect(response.status).toBe(404);
      expect(await response.text()).toBe("Not found");
    });

    it("should handle known routes", async () => {
      const request = new Request("https://example.com/code");
      const url = new URL("https://example.com/code");

      const response = await routeHandler.handleRoute(request, url, ["code"]);

      expect(response.status).toBe(200);
      expect(await response.text()).toBe("mock code");
    });
  });

  describe("Route Handlers", () => {
    describe("handleCodeRoute", () => {
      it("should return code with correct headers", async () => {
        const request = new Request("https://example.com/code");
        const url = new URL("https://example.com/code");

        const response = await routeHandler.handleRoute(request, url, ["code"]);

        expect(response.status).toBe(200);
        expect(await response.text()).toBe("mock code");
        expect(response.headers.get("Access-Control-Allow-Origin")).toBe("*");
        const md5 = await md5Promise;
        expect(response.headers.get("content_hash")).toBe(md5("mock code"));
      });
    });

    describe("handleSessionRoute", () => {
      it("should return session JSON", async () => {
        const request = new Request("https://example.com/session.json?room=test-space");
        const url = new URL("https://example.com/session.json?room=test-space");

        const response = await routeHandler.handleRoute(request, url, ["session.json"]);

        expect(response.status).toBe(200);
        expect(response.headers.get("Content-Type")).toBe("application/json; charset=UTF-8");
      });
    });

    describe("handleWebsocketRoute", () => {
      it("should handle websocket upgrade", async () => {
        const request = new Request("https://example.com/websocket", {
          headers: new Headers({ "Upgrade": "websocket" }),
        });
        const url = new URL("https://example.com/websocket");

        const response = await routeHandler.handleRoute(request, url, ["websocket"]);

        expect(response.status).toBe(101);
        expect(response.webSocket).toBeDefined();
        expect(response.webSocket).toHaveProperty("send");
        expect(response.webSocket).toHaveProperty("close");
        expect(mockCode.wsHandler?.handleWebsocketSession).toHaveBeenCalled();
      });

      it("should return 400 for non-websocket request", async () => {
        const request = new Request("https://example.com/websocket");
        const url = new URL("https://example.com/websocket");

        const response = await routeHandler.handleRoute(request, url, ["websocket"]);

        expect(response.status).toBe(400);
        expect(await response.text()).toBe("Expected websocket");
      });
    });

    describe("handleDefaultRoute", () => {
      it("should return HTML with correct replacements", async () => {
        const request = new Request("https://example.com");
        const url = new URL("https://example.com");

        const response = await routeHandler.handleRoute(request, url, [""]);

        expect(response.status).toBe(200);
        const responseText = await response.text();

        // Check for importmap replacement
        expect(responseText).toContain(
          `<script type="importmap">${JSON.stringify(importMap)}</script>`,
        );

        // Check for CSS link replacement
        expect(responseText).toContain(
          `<link rel="preload" href="/live/test-space/index.css" as="style" />`,
        );

        // Check for embed div replacement
        expect(responseText).toContain(`<div id="embed">mock html</div>`);
      });
    });

    describe("handleJsRoute", () => {
      it("should return transpiled JS with correct headers", async () => {
        const request = new Request("https://example.com/index.js");
        const url = new URL("https://example.com/index.js");

        const response = await routeHandler.handleRoute(request, url, ["index.js"]);

        expect(response.status).toBe(200);
        expect(response.headers.get("Content-Type")).toBe("application/javascript; charset=UTF-8");
      });
    });

    describe("handleCssRoute", () => {
      it("should return CSS with correct headers", async () => {
        const request = new Request("https://example.com/index.css");
        const url = new URL("https://example.com/index.css");

        const response = await routeHandler.handleRoute(request, url, ["index.css"]);

        expect(response.status).toBe(200);
        expect(await response.text()).toBe("mock css");
        expect(response.headers.get("Content-Type")).toBe("text/css; charset=UTF-8");
      });
    });
  });
});
