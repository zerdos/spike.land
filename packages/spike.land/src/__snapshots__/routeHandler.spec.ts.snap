// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RouteHandler > Route Handlers > handleDefaultRoute > should return HTML with correct replacements 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <base href="/" />
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>spike.land</title>
    <link
      rel="preload"
      href="https://fonts.googleapis.com/css2?family=Roboto+Flex:opsz,wght@8..144,100..1000&display=swap"
      as="style"
    />
    <link
      href="https://fonts.googleapis.com/css2?family=Roboto+Flex:opsz,wght@8..144,100..1000&display=swap"
      rel="stylesheet"
    />

    <script type="importmap">
      {
        "imports": {
          "@/": "/@/",
          "@emotion/react/jsx-runtime": "/emotionJsxRuntime.mjs",
          "@emotion/react/jsx-dev-runtime": "/emotionJsxRuntime.mjs",
          "@emotion/styled": "/emotionStyled.mjs", 
          "react/jsx-runtime": "/jsx.mjs",
          "react-dom/server": "/reactDomServer.mjs",
          "react-dom/client": "/reactDomClient.mjs",
          "@emotion/react": "/emotion.mjs",
          "react": "/reactMod.mjs",
          "framer-motion": "/@/workers/framer-motion.mjs",
          "react-dom": "/reactDom.mjs"
        }
      }
    </script>
  </head>
  <style>
  mock css
  </style>

  <body>
    
    <div id="embed">mock html</div>
    <script type="module" src="/start.mjs"></script>
  </body>
</html>
"
`;
